lib/api/core/validation/index.js

```diff
@@ -1,690 +1,680 @@
-        });
+  /**
+   * @param {string} indexName
+   * @param {string} collectionName
+   * @param {CollectionSpecification} collectionSpec
+   * @param {boolean} [verboseErrors]
+   * @returns {Promise<boolean>}
+   */
+  isValidSpecification(indexName, collectionName, collectionSpec, verboseErrors = false) {
+    // We make a deep clone to avoid side effects
+    const specification = _.cloneDeep(collectionSpec);
+
+    return new Bluebird(resolve => {
+      this
+        .curateCollectionSpecification(indexName, collectionName, specification, true, verboseErrors)
+        .then(result => {
+          if (verboseErrors && result.isValid === false) {
+            return resolve(result);
+          }
+
+          return resolve({isValid: true});
+        })
+        // we do not want to reject since this method goal is
+        // to know what is going wrong with the given spec
+        .catch(error => resolve({isValid: false, errors: [error]}));
     });
-};
+  }
 
-/**
- * @param {string} indexName
- * @param {string} collectionName
- * @param {CollectionSpecification} collectionSpec
- * @param {boolean} verboseErrors
- * @returns {Promise<boolean>}
- */
-Validation.prototype.isValidSpecification = function validationIsValidSpecification (indexName, collectionName, collectionSpec, verboseErrors) {
-  // We make a deep clone to avoid side effects
-  const specification = _.cloneDeep(collectionSpec);
-  verboseErrors = verboseErrors || false;
-
-  return new Promise (resolve => {
-    this
-      .curateCollectionSpecification(indexName, collectionName, specification, true, verboseErrors)
-      .then(result => {
+  /**
+   * @param {string} indexName
+   * @param {string} collectionName
+   * @param {CollectionSpecification} collectionSpec
+   * @param {boolean} [dryRun]
+   * @param {boolean} [verboseErrors]
+   * @returns {CollectionSpecification}
+   * @rejects InternalError
+   */
+  curateCollectionSpecification(indexName, collectionName, collectionSpec, dryRun = false, verboseErrors = false) {
+    let errorMessage = '';
+
+    return new Bluebird((resolve, reject) => {
+      const
+        treatedSpecification = {
+          strict: collectionSpec.strict || false,
+          fields: {},
+          validators: null
+        };
+
+      if (!checkAllowedProperties(collectionSpec, ['strict', 'fields', 'validators'])) {
+        errorMessage = `${indexName}.${collectionName}: the collection specification has invalid properties.`;
         if (verboseErrors) {
+          return resolve({isValid: false, errors: [errorMessage]});
+        }
+        return reject(new KuzzleInternalError(errorMessage));
+      }
+
+      if (collectionSpec.fields) {
+        try {
+          const result = this.structureCollectionValidation(collectionSpec, indexName, collectionName, verboseErrors);
           if (result.isValid === false) {
-            return resolve(result);
+            if (verboseErrors) {
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png 'Severity: CRITICAL') Refactor this code to not nest more than 3 if/for/while/switch/try statements. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3ANestedIfDepth)